envoy/http/filter.h

```diff
@@ -399,7 +399,7 @@ class StreamFilterCallbacks {
    * This is a helper to get the route's per-filter config if it exists, otherwise the virtual
    * host's. Or nullptr if none of them exist.
    */
-  virtual const Router::RouteSpecificFilterConfig* mostSpecificPerFilterConfig() const PURE;
```

Comment:
Why don't we just return a shared_ptr here directly?

Reply:
Almost no filter use this method directly. most filters use a utility function to get the typed config. So I create a variant of origin utility function to get a typed shared pointer. For the filters that need not keep the ownership of config, they can use origin utility function directly.